### PR TITLE
Move signature tests from cambridge suite to component tests

### PR DESCRIPTION
**Pull request by T-Gro**
This is a movement of a suite testing a full signature roundtrip:
- Start with an .fs/.fsx implementation file
- Autogenerate a signature for it
- Compile the original file together with the signature

When doing this, two new issues regarding signature generation were opened.
The tests used to work by invoking fsc.exe, is now done in memory using component tests.
Also, the verification in the older suite most likely did not catch compiler issues, because I had to fight many.
This is now captured in the test assert, that compilation must succeed.

### CODE REVIEW

File: tests/FSharp.Compiler.ComponentTests/FSharp.Compiler.ComponentTests.fsproj [hunk 1]
```diff
@@ -207,6 +207,7 @@
     <Compile Include="Signatures\RecordTests.fs" />
     <Compile Include="Signatures\ArrayTests.fs" />
     <Compile Include="Signatures\TypeTests.fs" />
+	<Compile Include="Signatures\SigGenerationRoundTripTests.fs" />
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/FSharp.Compiler.ComponentTests.fsproj:210]: Spacing here

File: tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access-minimal-repro.fsx [hunk 2]
```diff
@@ -0,0 +1,8 @@
+module Core_access
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access-minimal-repro.fsx:1]: Maybe also put into the test module?

File: tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access-minimal-repro.fsx [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+    
+[<Sealed>]
+type MyClassPropertiyGetters =
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access-minimal-repro.fsx:4]: Typo

File: tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access-minimal-repro.fsx (hunk 2 above)
Comment (psfinaki) [reply tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access-minimal-repro.fsx:1]: (same in some other files)

File: tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access.fsx [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+                    {
+                        Thing: HelperModule.Data
+                    }
+                    
+            let getInt (data:Data): int = HelperModule.handle data.Thing               
+
+
+    (*--------------------*)  
+
+#if TESTS_AS_APP
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access.fsx:274]: Just double-checking if the constants are still applicable / still apply after all the file migration?

File: tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/array.fsx [hunk 5]
```diff
@@ -0,0 +1,1151 @@
...
+                    MarkerStyle.Star10
+                    MarkerStyle.Triangle
+                    MarkerStyle.Triangle1
+                |] 
+
+    do check "bug872632" Foo.x.Length 8
+
+module CheckUnionTypesAreSealed =
+    open System
+#if NETCOREAPP
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/array.fsx:1048]: Still required here?

**Review by psfinaki — COMMENTED** [review]
Good stuff, great job. Put a few small remarks, of course, there is probably a lot that can be improved (e.g. inconsistent casing) but this is not the point of the PR.

Food for thought - I'm currently splitting Editor.Tests into a separate project, have you considered splitting this into a separate project as well? "Component tests" are quite bloated, to the level the name of the project doesn't tell much anymore.

File: tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access.fsx (hunk 4 above)
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access.fsx:274]: Was just a file move, with the **intention to move as many things as is**, without a change.

File: tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access-minimal-repro.fsx (hunk 2 above)
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access-minimal-repro.fsx:1]: There is no standard, the test files simply carried whichever module they had before (minimizing number of changes) and this is a subset of the big access.fsx to demonstrate an issue with sig file gen.

File: tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access.fsx (hunk 4 above)
Comment (psfinaki) [reply tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access.fsx:274]: Well I just don't know how and where this constant is defined. Sometimes constants are defined in project files, so when files are moved from one project to another project, some constants should be copied. But I don't know if that's the case here, hence double-checking :)

File: tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/array.fsx (hunk 5 above)
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/array.fsx:1048]: The code has been in general kept unless things were failing.
This is a harmless polyfill , keeping it.

File: tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access.fsx (hunk 4 above)
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access.fsx:274]: The code driving the suite is where all the important stuff happens.
Defines are part of it:

```
    Fsi generatedSignature    
    |> withAdditionalSourceFile (FsSource implContents)
    |> withLangVersionPreview
    |> withDefines ["TESTS_AS_APP";"COMPILED"]
```

**Comment by T-Gro** [comment]
> Good stuff, great job. Put a few small remarks, of course, there is probably a lot that can be improved (e.g. inconsistent casing) but this is not the point of the PR.
> 
> Food for thought - I'm currently splitting Editor.Tests into a separate project, have you considered splitting this into a separate project as well? "Component tests" are quite bloated, to the level the name of the project doesn't tell much anymore.

Personal opinion - 
Unless it will be crystal clear which test project exists for which compiler feature, I am strongly against new test projects.
It then leads to the situation where "signature file related tests" are spread across multiple projects, without a clear idea where new test code should go (because the established frameworks and helpers provide slightly different functionalities). And we did see incoming commits even to some of the more older test projects, yes.

For a very independent set of features, say "Tests for VS tooling", it might make sense **if** all existing tests are moved to a single place as well (even if just copy pasted using their existing mechanisms)

**Review by vzarytovskii — APPROVED** [review]

File: tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access.fsx (hunk 4 above)
Comment (psfinaki) [reply tests/FSharp.Compiler.ComponentTests/Signatures/TestCasesForGenerationRoundTrip/access.fsx:274]: Ah okay I see now, thanks, it's fine then

**Review by psfinaki — APPROVED** [review]

**Pull request merged**